MapboxCoreNavigation/RouteController.swift

```diff
@@ -648,6 +758,24 @@ extension RouteController: CLLocationManagerDelegate {
         return true
     }
     
+    /**
+     Given a user's current location, an intersection that contains a tunnel road class and the current distance to the intersection provided,
+     returns a Boolean whether they are within the minimum radius of a tunnel entrance.
+     */
+    public func userWithinTunnelEntranceRadius(at location: CLLocation, intersection: Intersection?, distance: CLLocationDistance?) -> Bool {
+        // Ensure the upcoming intersection is a tunnel intersection.
+        guard let upcomingIntersection = intersection,
+              let roadClasses = upcomingIntersection.outletRoadClasses, roadClasses.contains(.tunnel),
+              (location.speed >= RouteControllerMinimumSpeedAtTunnelEntranceRadius || !location.isQualified) else {
+                return false
+        }
+        
+        // Distance to the upcoming tunnel entrance
+        guard let distanceToTunnelEntrance = distance else { return false }
```

Comment:
This line is interesting. Why not make `distance` a non-optional?

Reply:
@JThramer that is intentional. `routeProgress.currentLegProgress.currentStepProgress.userDistanceToUpcomingIntersection` happens to be an optional value. Rather than using a default value, I'm just going to exit the function if the value of `userDistanceToUpcomingIntersection` happens to be `nil`.

Reply:
Yeah, that's understood, the crux of what I'm getting at here is that "when it comes to functions that answer a question, it's better to resolve optionality of data at the call-site rather than burying that logic in the function."

Sometimes it's okay to handle optionals inside a function, for example when the possibility of non-existence is part of the "semantic domain" of the function itself -- but I would argue that this is not one of those cases, as "the edge-case of the device not having a CLLocation yet" is not within the problem domain of "Is this location within the tunnel-distance threshold?"

So applying this principle, the function would be written more like this: 


```swift
func shouldEnableTunnelAnimation(at location: CLLocation, for manager: CLLocationManager, intersection: Intersection?) -> Bool {
    guard let currentIntersection = intersection else { return false }
    
    let stepProgress = routeProgress.currentLegProgress.currentStepProgress
    
    guard let classes = currentIntersection.outletRoadClasses,
          let upcoming = stepProgress.upcomingIntersection,
          let upcomingDistance = stepProgress.userDistanceToUpcomingIntersection else { return false }
    
    let isWithinTunnelRadius = userWithinTunnelEntranceRadius(at: location, intersection: upcoming, distance: upcomingDistance)
    let isInsideTunnel = (classes.contains(.tunnel) && (manager is NavigationLocationManager && !location.isQualified))
    
    return isWithinTunnelRadius || isInsideTunnel
}
```

Doesn't that seem cleaner?

Reply:
@JThramer I agree we can definitely it's okay to handle optionals inside a function. I have 2 goals here. 

Firstly, make my function very loosely-coupled and testable. So with in mind, you will notice the conditional assignment of the `userDistanceToUpcomingIntersection` makes our function very tightly coupled and our test brittle. 

Secondly, my final effort in this PR will be tripping this tunnel work entirely out of the `RouteController` class. My tunnel detector class (coming soon...) should not even know of `routeProgress.currentLegProgress.currentStepProgress`. 

My goal is to make my functions be simple and easy to plug & play.

----
Further analysis:

Per existing logic, we conditionally set `userDistanceToUpcomingIntersection` and I would rather not make this function very dependable on the routeController properties such as `routeProgress`.  We will be ending up making this work so tightly coupled to the RouteController. Which is what we really want to avoid. 

Makes sense?

Code Reference: (Where `userDistanceToUpcomingIntersection` is set.)

https://github.com/mapbox/mapbox-navigation-ios/blob/1d6f00b6d7e604dff75a2e57f9e0fab21d2b9ff3/MapboxCoreNavigation/RouteController.swift#L955